games/static/games/proto.js

```diff
@@ -0,0 +1,255 @@
+    constructor(name, health, square, lineWidth) {
+        this.name = name
+        this.imageCssClass = ['dead' , 'invader', 'intruder']
+        this.health = health
+        this.square = square
+        this.lineWidth = lineWidth
+        this.type = 'alien'
+    }
+    getCssClass() {
+        return this.imageCssClass[this.health]
+    }
+    moveTo(square) {
+        square.setCharacter(this)
+    }
+    isDead() {
+        return this.health === 0
+    }
+}
+
+function createInvader(health ,indexOnBoard, lineWidth) {
+    let alien = new Alien('invader',health , indexOnBoard, lineWidth)
+    return alien
+}
+
+function createIntruder(indexOnBoard, lineWidth) {
+    return new Alien('intruder', 2, indexOnBoard, lineWidth)
+}
+
+class Board {
+    constructor(parent, width, setup) {
+        this.parent = parent
+        this.width = width
+        this.aliens = []
+        this.squares = []
+        this.setup = setup
+        this.setupObject = setup.getAlienLocations()
+        this.alienLocations = Object.keys(this.setupObject).map(Number)
+        this.alienHealthLevels = Object.values(this.setupObject)
+        this.direction = 1
+        this.defenderLocation = this.defenderStartPosition()
+
+    }
+    createChildSquareAddToSquares(index) {
+        let div = document.createElement('div')
+        this.parent.appendChild(div)
+        const borders = []
+        if (index >= this.bottomRow()) {
+            borders.push('ground')
+        } else if (index % this.width === 0) {
+            borders.push('left')
+        } else if (index % this.width === this.width-1) {
+            borders.push('right')
+        }
+        this.squares.push(new Square(index, div, this.width, borders))
+    }
+    bottomRow() {
+        return (this.width * this.width - this.width)
+    }
+    defenderStartPosition() {
+        return this.width * this.width - (this.width + Math.floor(this.width / 2) + 1)
+    }
+
+    draw() {
+        console.log(this.alienHealthLevels)
+        for (let i = 0; i < this.width * this.width; i++) {
+            this.createChildSquareAddToSquares(i)
+            if (this.alienLocations.includes(i)) {
+                let health = this.alienHealthLevels.shift()
+                let invader = createInvader(health, this.squares[i].index, this.width)
+                this.squares[i].setCharacter(invader)
+            }
+            if (i >= this.bottomRow()) {
+                this.squares[i].setGround()
+            }
+        }
+        let defender = new Defender('defender', this.width)
+        this.squares[this.defenderLocation].setCharacter(defender)
+    }
+
+    
+    moveAliens() {
+        this.determineDirection()
+        for (let i = 0; i < this.squares.length - 1; i++) {
+            if(this.squares[i].hasAlien()) {
+                let c = this.squares[i].pickCharacterUp()
+                this.aliens.push(c)
+            }
+        }
+        for (let i = 0; i < this.alienLocations.length; i++) {
```

Comment:
also here: i would calculate a new array of alien locations and only at the end mutate the class member:
```javascript
this.alienLocations = this.alienLocations.map(loc=>loc+this.direction)
```
and i wonder if our original plan was so bad:
```javascript
this.alienLocations = this.aliens.map(alien=>alien.locationOnBoard+this.direction)
```